SL/Controller/ZUGFeRD.pm

```diff
@@ -77,72 +97,174 @@ sub action_import_zugferd {
 
-  my $invnumber = $dom->findnodes('//rsm:ExchangedDocument/ram:ID')->string_value;
+  if ( $taxnumber ) {
+    $vendor = find_vendor_by_taxnumber($taxnumber);
+  }
 
-  # parse dates to kivi if set/valid
-  my ($transdate, $duedate, $dt_to_kivi, $due_dt_to_kivi);
-  $transdate = $dom->findnodes('//ram:IssueDateTime')->string_value;
-  $duedate   = $dom->findnodes('//ram:DueDateDateTime')->string_value;
-  $transdate =~ s/^\s+|\s+$//g;
-  $duedate   =~ s/^\s+|\s+$//g;
-
-  if ($transdate =~ /^[0-9]{8}$/) {
-    $dt_to_kivi = DateTime->new(year  => substr($transdate,0,4),
-                                month => substr ($transdate,4,2),
-                                day   => substr($transdate,6,2))->to_kivitendo;
+  if (ref $vendor eq 'SL::DB::Vendor') { return $vendor; }
+
+  return undef;
+}
+
+sub action_import_zugferd {
+  my ($self, %params) = @_;
+
+  my %res;          # result data structure returned by SL::ZUGFeRD->extract_from_{pdf,xml}()
+  my $parser;       # SL::XMLInvoice object created by SL::ZUGFeRD->extract_from_{pdf,xml}()
+  my $dom;          # DOM object for parsed XML data
+  my $template_ap;  # SL::DB::RecordTemplate object
+  my $vendor;       # SL::DB::Vendor object
+
+  my $ibanmessage;  # Message to display if vendor's database and invoice IBANs don't match up
+
+  die t8("missing file for action import") unless $::form->{file};
+  die t8("can only parse a pdf or xml file")      unless $::form->{file} =~ m/^%PDF|<\?xml/;
+
+  if ( $::form->{file} =~ m/^%PDF/ ) {
+    %res = %{SL::ZUGFeRD->extract_from_pdf($::form->{file})}
+  } else {
+    %res = %{SL::ZUGFeRD->extract_from_xml($::form->{file})};
   }
-  if ($duedate =~ /^[0-9]{8}$/) {
-    $due_dt_to_kivi = DateTime->new(year  => substr($duedate,0,4),
-                                    month => substr ($duedate,4,2),
-                                    day   => substr($duedate,6,2))->to_kivitendo;
+
+  if ($res{'result'} != SL::ZUGFeRD::RES_OK()) {
+    # An error occurred; log message from parser:
+    $::lxdebug->message(LXDebug::DEBUG1(), "Could not extract ZUGFeRD data, error message: " . $res{'message'});
+    die(t8("Could not extract Factur-X/ZUGFeRD data, data and error message:") . " $res{'message'}");
   }
 
-  my $type = $dom->findnodes('//rsm:ExchangedDocument/ram:TypeCode')->string_value;
+  $parser = $res{'invoice_xml'};
+
+  # Shouldn't be neccessary with SL::XMLInvoice doing the heavy lifting, but
+  # let's grab it, just in case.
+  $dom  = $parser->{dom};
+
+  my %metadata = %{$parser->metadata};
+  my @items = @{$parser->items};
+
+  my $iban = $metadata{'iban'};
+  my $invnumber = $metadata{'invnumber'};
+
+  if ( ! ($metadata{'ustid'} or $metadata{'taxnumber'}) ) {
+    die t8("Cannot process this invoice: neither VAT ID nor tax ID present.");
+    }
+
+  $vendor = find_vendor($metadata{'ustid'}, $metadata{'taxnumber'});
+
+  die t8("Please add a valid VAT ID or tax number for this vendor: #1", $metadata{'vendor_name'}) unless $vendor;
```

Comment:
Die Fehlermeldung ist zumindest irreführend — sie wird nämlich auch dann angezeigt, wenn es den Lieferanten noch gar nicht gibt. Besser wäre es, das in der Fehlermeldung auch explizit als Möglichkeit zu erwähnen. Beispiel:

> No vendor was found when looking for VAT ID or tax number <Nummer> (name in invoice: <Name>). The vendor might not exist yet, or if it does, the VAT ID or tax number might not be set.

Reply:
Repariert.